File: src/metropulse/domain/journey.py
"""Domain values for journey planning."""

from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime


@dataclass(frozen=True, slots=True)
class JourneyStop:
    """A station reference within a journey plan."""

    stop_id: str
    name: str


@dataclass(frozen=True, slots=True)
class RideLeg:
    """One continuous ride on a single line/direction.

    Carries exactly the identifiers (route_id, direction_id, boarding and
    alighting stops) a coach-recommendation or crowd-prediction engine needs
    to annotate the leg later — extension happens by composition, not by
    changing this type.
    """

    board: JourneyStop
    alight: JourneyStop
    route_id: str
    route_short_name: str | None
    route_long_name: str | None
    route_color: str | None
    direction_id: int | None
    platform_hint: str
    stations: tuple[JourneyStop, ...]
    ride_seconds: float
    wait_seconds: float

    @property
    def kind(self) -> str:
        """Leg discriminator."""
        return "ride"


@dataclass(frozen=True, slots=True)
class WalkLeg:
    """A walking transfer between two nearby stations."""

    board: JourneyStop
    alight: JourneyStop
    distance_m: float
    walk_seconds: float

    @property
    def kind(self) -> str:
        """Leg discriminator."""
        return "walk"


JourneyLeg = RideLeg | WalkLeg


@dataclass(frozen=True, slots=True)
class JourneyPlan:
    """A complete origin-to-destination plan."""

    origin: JourneyStop
    destination: JourneyStop
    departure_at: datetime
    expected_arrival_at: datetime
    expected_travel_seconds: float
    interchange_count: int
    walking_distance_m: float
    legs: tuple[JourneyLeg, ...]

    @property
    def remaining_stations(self) -> tuple[JourneyStop, ...]:
        """Every station passed after boarding, across all ride legs."""
        stations: list[JourneyStop] = []
        for leg in self.legs:
            if isinstance(leg, RideLeg):
                stations.extend(leg.stations[1:])
        return tuple(stations)

    @property
    def interchange_stops(self) -> tuple[JourneyStop, ...]:
        """The stations where the rider changes lines (alighting points)."""
        rides = [leg for leg in self.legs if isinstance(leg, RideLeg)]
        return tuple(leg.alight for leg in rides[:-1])

File: app/lib/domain/models/journey.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'journey.freezed.dart';
part 'journey.g.dart';

/// A stop within a journey plan.
@freezed
class JourneyStop with _$JourneyStop {
  const factory JourneyStop({required String stopId, required String name}) =
      _JourneyStop;

  factory JourneyStop.fromJson(Map<String, dynamic> json) =>
      _$JourneyStopFromJson(json);
}

/// One ride or walk leg of a plan.
@freezed
class JourneyLeg with _$JourneyLeg {
  const JourneyLeg._();

  const factory JourneyLeg({
    required String kind,
    required JourneyStop board,
    required JourneyStop alight,
    required double seconds,
    String? routeId,
    String? routeLongName,
    String? routeColor,
    int? directionId,
    String? platformHint,
    double? waitSeconds,
    List<JourneyStop>? stations,
    double? distanceM,
  }) = _JourneyLeg;

  factory JourneyLeg.fromJson(Map<String, dynamic> json) =>
      _$JourneyLegFromJson(json);

  bool get isRide => kind == 'ride';
}

/// A complete origin-to-destination plan.
@freezed
class JourneyPlan with _$JourneyPlan {
  const factory JourneyPlan({
    required JourneyStop origin,
    required JourneyStop destination,
    required DateTime departureAt,
    required DateTime expectedArrivalAt,
    required double expectedTravelSeconds,
    required int interchangeCount,
    required List<String> interchangeStopIds,
    required double walkingDistanceM,
    required List<JourneyStop> remainingStations,
    required List<JourneyLeg> legs,
  }) = _JourneyPlan;

  factory JourneyPlan.fromJson(Map<String, dynamic> json) =>
      _$JourneyPlanFromJson(json);
}

/// A tracked journey (the server-side session).
@freezed
class Journey with _$Journey {
  const factory Journey({
    required int id,
    required String originStopId,
    required String destinationStopId,
    required String status,
    required DateTime startedAt,
    String? routeId,
    String? vehicleId,
    DateTime? endedAt,
  }) = _Journey;

  factory Journey.fromJson(Map<String, dynamic> json) => _$JourneyFromJson(json);
}
